### PR TITLE
End to end example for Admin and web API sites

### PR DESCRIPTION
**Pull request by s3364859**
@6nop @s3494110 @s3482247 

In this PR:
- Using the Test table, I've whipped up a bunch of scaffolding around that entity
- Added in an admin controller
- Added in admin model
- Added in appropriate CRUD operations in the DAL and also in views
- Updated web API project with examples based on the Test entity

Not done:
- Paging
- Sorting

I will get onto these later on.

goose-parade

Not sure how marching poultry relates to this PR - pencil this one in yourselves!

### CODE REVIEW

File: CPT331.Data.Migration/Migrations/M0001.sql [hunk 1]
```diff
@@ -1,4 +1,42 @@
-﻿CREATE PROCEDURE dbo.spGetTest
+﻿CREATE PROCEDURE dbo.spAddTest
```
Comment (s3364859) [inline CPT331.Data.Migration/Migrations/M0001.sql]: Added in sprocs for Add, Delete, Update, Get, and GetByID. Alphabetised the file too.

File: CPT331.Data/Test.cs [hunk 2]
```diff
@@ -11,8 +11,8 @@ public class Test
 		public Test(int id, string value0, string value1)
 		{
 			_id = id;
-			_value1 = value0;
-			_value0 = value1;
+			_value0 = value0;
+			_value1 = value1;
```
Comment (s3364859) [inline CPT331.Data/Test.cs]: This caused some headaches!

File: CPT331.Data/TestRepository.cs [hunk 3]
```diff
@@ -13,19 +14,65 @@ namespace CPT331.Data
 {
 	public static class TestRepository
 	{
-		public static List<Test> GetTests()
+		public static int AddTest(string value0, string value1)
```
Comment (s3364859) [inline CPT331.Data/TestRepository.cs]: Corresponding CRUD operations (alphabetised too).

File: CPT331.Data/TestRepository.cs [hunk 4]
```diff
@@ -13,19 +14,65 @@ namespace CPT331.Data
...
+		{
+			int id = 0;
+
+			using (SqlConnection sqlConnection = SqlConnectionFactory.NewSqlConnetion())
+			{
+				id = (int)SqlMapper
```
Comment (s3364859) [inline CPT331.Data/TestRepository.cs]: Return the ID of the newly added item.

File: CPT331.Web.Tests/Controllers/HomeControllerTest.cs [hunk 5]
```diff
@@ -19,7 +19,7 @@ public void Index()
 			HomeController controller = new HomeController();
 
 			// Act
-			ViewResult result = controller.Index() as ViewResult;
+			ViewResult result = controller.Home() as ViewResult;
```
Comment (s3364859) [inline CPT331.Web.Tests/Controllers/HomeControllerTest.cs]: Change 'Index' to 'Home' - they mean different things to me, an index page is not always a home page etc.

File: CPT331.Web/Controllers/AdminController.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+using CPT331.Web.Models.Admin;
+using System.Collections;
+using System.Collections.Generic;
+
+#endregion
+
+namespace CPT331.Web.Controllers
+{
+    public class AdminController : Controller
```
Comment (s3364859) [inline CPT331.Web/Controllers/AdminController.cs]: New AdminController type.

File: CPT331.Web/Controllers/AdminController.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    {		
+		[AcceptVerbs(HttpVerbs.Get)]
+        public ActionResult Home()
+        {
+            return View();
+        }
+
+		[AcceptVerbs(HttpVerbs.Get)]
+		public ActionResult NewTest()
```
Comment (s3364859) [inline CPT331.Web/Controllers/AdminController.cs]: Add a new Test.

File: CPT331.Web/Controllers/AdminController.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+			else
+			{
+				actionResult = View(testModel);
+			}
+
+			return actionResult;
+		}
+
+		[AcceptVerbs(HttpVerbs.Get)]
+        public ActionResult Test(int id)
```
Comment (s3364859) [inline CPT331.Web/Controllers/AdminController.cs]: Get a single Test by ID.

File: CPT331.Web/Controllers/AdminController.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            else
+            {
+                actionResult = View(testModel);
+            }
+
+            return actionResult;
+        }
+
+		[AcceptVerbs(HttpVerbs.Get)]
+		public ActionResult Tests()
```
Comment (s3364859) [inline CPT331.Web/Controllers/AdminController.cs]: Get all the Test entities.

File: CPT331.Web/Models/Admin/TestModel.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+using System;
+
+#endregion
+
+namespace CPT331.Web.Models.Admin
+{
+    public class TestModel
+    {
+        public TestModel()
```
Comment (s3364859) [inline CPT331.Web/Models/Admin/TestModel.cs]: This empty constructor is unfortunately required by MVC.

File: CPT331.WebAPI/Controllers/TestController.cs [hunk 11]
```diff
@@ -1,39 +1,64 @@
...
-        public void Put(int id, [FromBody]string value)
-        {
-        }
-
-        // DELETE api/<controller>/5
-        public void Delete(int id)
-        {
-        }
+		// GET api/<controller>
+		public IEnumerable<TestModel> Get()
```
Comment (s3364859) [inline CPT331.WebAPI/Controllers/TestController.cs]: Get = return a list of T.

File: CPT331.WebAPI/Controllers/TestController.cs [hunk 12]
```diff
@@ -1,39 +1,64 @@
...
+		{
+			return TestRepository
+				.GetTests()
+				.Select(m => new TestModel(m.ID, m.Value0, m.Value1));
+		}
+
+		// GET api/<controller>/5
+		public TestModel Get(int id)
```
Comment (s3364859) [inline CPT331.WebAPI/Controllers/TestController.cs]: Get() = return a single T.

File: CPT331.WebAPI/Controllers/TestController.cs [hunk 13]
```diff
@@ -1,39 +1,64 @@
...
+			if (test != null)
+			{
+				testModel = new TestModel(test.ID, test.Value0, test.Value1);
+			}
+
+			return testModel;
+		}
+
+		// POST api/<controller>
+		public void Post(TestModel testModel)
```
Comment (s3364859) [inline CPT331.WebAPI/Controllers/TestController.cs]: Post = make a new T.

File: CPT331.WebAPI/Controllers/TestController.cs [hunk 14]
```diff
@@ -1,39 +1,64 @@
...
+		{
+			if (ModelState.IsValid == true)
+			{
+				TestRepository.AddTest(testModel.Value0, testModel.Value1);
+			}
+		}
+
+		// PUT api/<controller>/5
+		public void Put(TestModel testModel)
```
Comment (s3364859) [inline CPT331.WebAPI/Controllers/TestController.cs]: Put = update an existing T.

File: CPT331.WebAPI/Controllers/TestController.cs [hunk 15]
```diff
@@ -1,39 +1,64 @@
...
+		{
+			if (ModelState.IsValid == true)
+			{
+				TestRepository.UpdateTest(testModel.ID, testModel.Value0, testModel.Value1);
+			}
+		}
+
+		// DELETE api/<controller>/5
+		public void Delete(int id)
```
Comment (s3364859) [inline CPT331.WebAPI/Controllers/TestController.cs]: DELETE(n) -> remove corresponding T.

**Comment by 6nop** [comment]
I'm happy with the changes & will complete the merge 👍

**Pull request merged**